libzen-compositor/keyboard.c

```diff
@@ -0,0 +1,244 @@
+    zen_log("keyboard: failed to get XKB keymap as string\n");
+    goto out_keymap;
+  }
+
+  keyboard->keymap_size = strlen(keymap_string) + 1;
+  keyboard->keymap_format = ZGN_KEYBOARD_KEYMAP_FORMAT_XKB_V1;
+  keyboard->keymap_fd =
+      zen_util_create_shared_file(keyboard->keymap_size, keymap_string);
+  if (keyboard->keymap_fd < 0) {
+    zen_log("keyboard: failed to create a keymap file\n");
+    goto out_file;
+  }
+
+  free(keymap_string);
+
+  xkb_keymap_unref(keymap);
+
+  xkb_context_unref(context);
+
+  return 0;
+
+out_file:
+  free(keymap_string);
+
+out_keymap:
+  xkb_keymap_unref(keymap);
+
+out_context:
+  xkb_context_unref(context);
+
+out:
+  return -1;
+}
+
+WL_EXPORT void
+zen_keyboard_set_focus(
+    struct zen_keyboard* keyboard, struct zen_virtual_object* virtual_object)
+{
+  struct zen_virtual_object* current_focus;
+
+  current_focus =
+      zen_weak_link_get_user_data(&keyboard->focus_virtual_object_link);
+
+  if (current_focus == virtual_object) return;
+
+  if (current_focus) zen_keyboard_leave(keyboard);
+  if (virtual_object) zen_keyboard_enter(keyboard, virtual_object);
+
+  if (virtual_object)
+    zen_weak_link_set(
+        &keyboard->focus_virtual_object_link, virtual_object->resource);
+  else
+    zen_weak_link_unset(&keyboard->focus_virtual_object_link);
+}
+
+WL_EXPORT void
+zen_keyboard_keymap(struct zen_keyboard* keyboard, struct wl_client* client)
+{
+  struct zen_keyboard_client* keyboard_client;
+  struct wl_resource* resource;
+
+  keyboard_client = zen_keyboard_client_find(client, keyboard);
+  if (keyboard_client == NULL) return;
+
+  wl_resource_for_each(resource, &keyboard_client->resource_list)
+  {
+    zgn_keyboard_send_keymap(resource, keyboard->keymap_format,
+        keyboard->keymap_fd, keyboard->keymap_size);
+  }
+}
+
+WL_EXPORT struct zen_keyboard*
+zen_keyboard_create(struct zen_seat* seat)
+{
+  struct zen_keyboard* keyboard;
+  int ret;
+
+  keyboard = zalloc(sizeof *keyboard);
+  if (keyboard == NULL) {
+    zen_log("keyboard: failed to allocate memory\n");
+    goto err;
+  }
+
+  ret = zen_keyboard_add_keymap(keyboard);
+  if (ret < 0) {
+    zen_log("keyboard: failed to add keymap\n");
+    goto err_keymap;
+  }
+  keyboard->seat = seat;
+  keyboard->grab = &keyboard->default_grab;
+  keyboard->default_grab.interface = &default_keyboard_grab_interface;
+  keyboard->default_grab.keyboard = keyboard;
+  zen_weak_link_init(&keyboard->focus_virtual_object_link);
+  wl_list_init(&keyboard->keyboard_client_list);
+  wl_signal_init(&keyboard->destroy_signal);
+  wl_array_init(&keyboard->keys);
+
+  return keyboard;
+
+err_keymap:
+  free(keyboard);
+
+err:
+  return NULL;
+}
+
+WL_EXPORT void
+zen_keyboard_destroy(struct zen_keyboard* keyboard)
+{
+  keyboard->grab->interface->cancel(keyboard->grab);
+  zen_weak_link_unset(&keyboard->focus_virtual_object_link);
+  wl_signal_emit(&keyboard->destroy_signal, NULL);
+  close(keyboard->keymap_fd);
+  wl_array_release(&keyboard->keys);
```

Comment:
```suggestion
  wl_array_release(&keyboard->keys); // TODO: release the contents of this array
```
めちゃ忘れそうだから付け足して欲しい。
https://github.com/zigen-project/zen/pull/22#discussion_r787923391